include/rdma/rdma_cma.h

```diff
@@ -538,6 +561,30 @@ int rdma_join_multicast(struct rdma_cm_id *id, struct sockaddr *addr,
 			void *context);
 
 /**
+ * rdma_join_multicast_ex - Joins a multicast group with options.
+ * @id: Communication identifier associated with the request.
+ * @mc_join_attr: Extensive struct containing multicast join parameters.
+ * @context: User-defined context associated with the join request.
+ * Description:
+ *   Joins a multicast group with options. Currently supporting MC join flags.
+ *   The QP will be attached based on the given join flag.
+ *   Join message will be sent according to the join flag.
+ * Notes:
+ *   Before joining a multicast group, the rdma_cm_id must be bound to
+ *   an RDMA device by calling rdma_bind_addr or rdma_resolve_addr.  Use of
+ *   rdma_resolve_addr requires the local routing tables to resolve the
+ *   multicast address to an RDMA device.  The user must call
+ *   rdma_leave_multicast to leave the multicast group and release any
+ *   multicast resources.  The context is returned to the user through
+ *   the private_data field in the rdma_cm_event.
+ * See also:
+ *   rdma_leave_multicast, rdma_bind_addr, rdma_resolve_addr, rdma_create_qp
+ */
+int rdma_join_multicast_ex(struct rdma_cm_id *id,
+			   struct rdma_cm_join_mc_attr_ex *mc_join_attr,
+			   void *context);
```

Comment:
We have some options here.  We could use rdma_set_option to specify more detailed/different multicast join operations, similar to setsockopt.  We could also define a new sockaddr type that allows us to specify more details -- this is more of a hack, but would be compatible.

If we have a new API, as long as we have a void * parameter, we can use a flag to indicate/change how that parameter should be interpreted by the library.  So there we can be extensible by adding a new flag parameter.

There's a trade off between being extensible and easy to use here.  Currently, it looks like the only goal is to support send-only joins.

As an option, we could define this:

rdma_join_multicast(struct rdma_cm_id *id, void *mc_addr, void *context, uint64_t flags);

Then the address doesn't need to be a sockaddr either, which would allow specifying the full set of IB multicast data.